src/test/java/dev/openfeature/sdk/providers/memory/InMemoryProviderTest.java

```diff
@@ -0,0 +1,88 @@
+package dev.openfeature.sdk.providers.memory;
+
+import com.google.common.collect.ImmutableMap;
+import dev.openfeature.sdk.Client;
+import dev.openfeature.sdk.OpenFeatureAPI;
+import dev.openfeature.sdk.Value;
+import lombok.SneakyThrows;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static dev.openfeature.sdk.Structure.mapToStructure;
+import static dev.openfeature.sdk.testutils.TestFlagsUtils.buildFlags;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class InMemoryProviderTest {
+
+    private static Client client;
+
+    private static final AtomicBoolean isProviderReady = new AtomicBoolean(false);
+
+    private static final AtomicInteger configurationChangedEventCount = new AtomicInteger(0);
+
+    @SneakyThrows
+    @BeforeAll
+    static void beforeAll() {
+        Map<String, Flag<?>> flags = buildFlags();
+        InMemoryProvider provider = new InMemoryProvider(flags);
+        OpenFeatureAPI.getInstance().setProvider(provider);
+
+        // TODO: setProvider with wait for init, pending https://github.com/open-feature/ofep/pull/80
+        Thread.sleep(500);
+
+        client = OpenFeatureAPI.getInstance().getClient();
+
+        client.onProviderReady(eventDetails -> isProviderReady.set(true));
+
+        client.onProviderConfigurationChanged(eventDetails -> configurationChangedEventCount.incrementAndGet());
+        provider.updateFlags(flags);
+        provider.updateFlag("addedFlag", Flag.builder()
+            .variant("on", true)
+            .variant("off", false)
+            .defaultVariant("on")
+            .build());
+    }
+
+    @SneakyThrows
+    @Test
+    void eventsTest() {
+        assertTrue(isProviderReady.get());
+        assertEquals(2, configurationChangedEventCount.get());
```

Comment:
There's an inherent race condition in this test. Here is an example failure: https://github.com/open-feature/java-sdk/actions/runs/5857702730/job/15882568501#step:6:150

This is because events run in a new thread and don't block, so it's possible for the `beforeAll` to fire the event, but the callback hasn't finished before this test runs.

I would recommend you use some sort of retry with a timeout, like Mochito's verify, like we do [here](https://github.com/open-feature/java-sdk/blob/b17a20e43fb2446da3f2b07d7a9d577a9b1dffea/src/test/java/dev/openfeature/sdk/EventsTest.java#L97). It should be true within a few ms.

Reply:
From inspecting, I think the issue was different here, you can update if I missed anything.
1. Handlers were set after setting the provider - see the [change](https://github.com/open-feature/java-sdk/pull/546/commits/41987b83b0657e3376980bc3ab44dab610dd18a0#diff-08b7a651a304317563611f9d4260b2f5e2ec554ba9fa6ae0e9a19a52efbd0d04).
This caused the possible race condition, as initialization was done in background, then calling the handler - which is set later.
2. "events run in a new thread and don't block" - I understand this is for provider ready event. But following the flow of _ProviderConfigurationChanged - provider.updateFlags(flags) -> emitProviderConfigurationChanged(details); -> this.onEmit.accept(this, event, details);_
I see this as direct non-blocking event call in same thread. Different provider implementation can to call the emit in a new thread.
3. I revisited the events in my last commit, you can re-review.

 As an early adopter of provider events with these issues, have some thoughts:
a. as (1) above was sort of implementation confusion, is there some way to enforce it  by the sdk ? should it be enforced ?
for example, when adding a new handler - check first if provider is set, otherwise throw exception ?
Should consider multiple / named providers though.
b. From EventProvider javadoc: "Please note that the SDK will automatically emit PROVIDER_READY" - I confused with it previously and manually emitted it without reading it possibly - can/should it be enforced by sdk ?
c. events are same for all providers in case of multiple named providers ?
d. I could call _OpenFeatureAPI.getInstance().onProviderReady_ and also _OpenFeatureAPI.getInstance().getClient().onProviderReady_. Why both are exposed ? Can you elaborate a bit on what is the difference between these ?
e. per (2) above, should the sdk emit these on the background ?

what do you think ?

Reply:
> Handlers were set after setting the provider - see the [change](https://github.com/open-feature/java-sdk/pull/546/commits/41987b83b0657e3376980bc3ab44dab610dd18a0#diff-08b7a651a304317563611f9d4260b2f5e2ec554ba9fa6ae0e9a19a52efbd0d04).
This caused the possible race condition, as initialization was done in background, then calling the handler - which is set later.

OK then this is working as expected. We want to support setting new providers, and we also want to allow users to add handlers at any time (before or after setting the provider) and for handlers to "move" to the new provider.

> 2.
> "events run in a new thread and don't block" - I understand this is for provider ready event. But following the flow of ProviderConfigurationChanged - provider.updateFlags(flags) -> emitProviderConfigurationChanged(details); -> this.onEmit.accept(this, event, details);
I see this as direct non-blocking event call in same thread. Different provider implementation can to call the emit in a new thread.

No matter where, "emit" happens, the SDK runs the _handler function_ in a new thread, see [here](https://github.com/open-feature/java-sdk/blob/b17a20e43fb2446da3f2b07d7a9d577a9b1dffea/src/main/java/dev/openfeature/sdk/EventSupport.java#L134).

> b. From EventProvider javadoc: "Please note that the SDK will automatically emit PROVIDER_READY" - I confused with it previously and manually emitted it without reading it possibly - can/should it be enforced by sdk ?

Even though the SDK automatically emits this event after `initialize()`, we still need to allow provider authors to sometimes emit it manually... for example, if the provider disconnects later and goes into ERROR state, it will then emit READY again when it reconnects. I'm not sure what we can do to prevent people from manually emitting it after init, which is not needed, as you say.

> c. events are same for all providers in case of multiple named providers ?
d. I could call OpenFeatureAPI.getInstance().onProviderReady and also OpenFeatureAPI.getInstance().getClient().onProviderReady. Why both are exposed ? Can you elaborate a bit on what is the difference between these ?

`Clients` are scoped to providers. The API is not. The API gets _all_ events. That's the difference between these. Client's are always only associated with a single provider, and evaluate flags from just that provider, and get events from just that provider.

> e. per (2) above, should the sdk emit these on the background ?

It's the provider that emits the events, which are "fire and forget" from it's perspective. They handlers associated with events run on a separate thread.

Reply:
> thank you for clarifying !
> I will try revisiting the tests for the issues.

Absolutely no problem! I think GH might be struggling right now though :grimacing: things seem to be queuing up and not completing :sweat_smile: